ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java

```diff
@@ -13,119 +13,447 @@
+            referencePartitionId, -1, goodUUID));
     // Extra characters after UUID
-    blobIdLikes.add(buildBlobIdLike(goodVersion, goodPartitionId, goodUUID.length(), goodUUID + "EXTRA"));
+    invalidBlobIdLikeList.add(
+        buildBlobIdLike(version, referenceFlag, referenceDatacenterId, referenceAccountId, referenceContainerId,
+            referencePartitionId, goodUUID.length(), goodUUID + "EXTRA"));
     // Invalid version number
-    blobIdLikes.add(buildBlobIdLike((short) (goodVersion + 1), goodPartitionId, goodUUID.length(), goodUUID));
-    // Empty blob ID
-    blobIdLikes.add("");
+    invalidBlobIdLikeList.add(
+        buildBlobIdLike((short) (BLOB_ID_V2 + 1), referenceFlag, referenceDatacenterId, referenceAccountId,
+            referenceContainerId, referencePartitionId, goodUUID.length(), goodUUID));
+    // Empty blobId
+    invalidBlobIdLikeList.add("");
     // short Blob ID
-    blobIdLikes.add("AA");
+    invalidBlobIdLikeList.add("AA");
 
-    for (String blobIdLike : blobIdLikes) {
+    for (String blobIdLike : invalidBlobIdLikeList) {
       try {
-        new BlobId(blobIdLike, clusterMap);
-        fail("Expected blob ID creation to fail with blob ID string " + blobIdLike);
+        BlobId.fromStringId(blobIdLike, referenceClusterMap);
+        fail("Expected blobId creation to fail with blobId string " + blobIdLike);
       } catch (Exception e) {
+        // expected
       }
     }
   }
 
   /**
-   * Build a string that resembles a blob ID, but with possibly invalid sections.
-   * @param version the version number to use in the blob ID
-   * @param partitionId the partition ID to use in the blob ID
-   * @param uuidLength the UUID length to use in the blob ID
-   * @param uuidLike the UUID to use in the blob ID
-   * @return a base-64 encoded {@link String} representing the blob ID
+   * Build a string that resembles a blobId, but with certain fields possibly set to legacy values.
+   * @param version The version number to be embedded in the blobId.
+   * @param flag The flag to be embedded in the blobId.
+   * @param datacenterId The datacenter id to be embedded in the blobId.
+   * @param accountId The account id to be embedded in the blobId.
+   * @param containerId The container id to be embedded in the blobId.
+   * @param partitionId The partition id to be embedded in the blobId.
+   * @param uuidLength The length of the uuid.
+   * @param uuidLike The UUID to be embedded in the blobId.
+   * @return a base-64 encoded {@link String} representing the blobId.
    */
-  private String buildBlobIdLike(short version, PartitionId partitionId, int uuidLength, String uuidLike) {
-    final int idLength = 2 + partitionId.getBytes().length + 4 + uuidLike.length();
-    ByteBuffer idBuf = ByteBuffer.allocate(idLength);
-    idBuf.putShort(version);
+  private String buildBlobIdLike(short version, Byte flag, Byte datacenterId, Short accountId, Short containerId,
```

Comment:
This is to mimic some fake ids, because blobid strings are not necessarily generated from `BlobId`, but can be maliciously made in various format. This is to ensure out deserialization is working well under malicious inputs.